File: frontend/src/types/etudiant.ts
/**
 * Types pour les entités du module Étudiant
 */

// ============ Etudiant ============
export interface Etudiant {
  id: number;
  matricule?: string;
  numero_carte?: string;
  ine?: string;
  nom: string;
  prenom: string;
  date_naissance?: string;
  lieu_naissance?: string;
  sexe?: string;
  nationalite?: string;
  photo_url?: string;
  email?: string;
  telephone?: string;
  telephone_urgence?: string;
  adresse?: string;
  ville?: string;
  pays?: string;
  nom_pere?: string;
  profession_pere?: string;
  nom_mere?: string;
  profession_mere?: string;
  personne_contact?: string;
  telephone_contact?: string;
  statut?: string;
  is_active?: boolean;
  created_at?: string;
  updated_at?: string;
  created_by?: string;
  last_modified_by?: string;
  last_modified_date?: string;
}

export type CreateEtudiant = Omit<Etudiant, 'id' | 'matricule' | 'created_at' | 'updated_at'>;

export type UpdateEtudiant = Partial<CreateEtudiant>;

export interface EtudiantWithDetails extends Etudiant {
  documents: DocumentEtudiant[];
  inscriptions: Inscription[];
}

export interface EtudiantStatistiques {
  total: number;
  par_statut: Record<string, number>;
  par_sexe: Record<string, number>;
}

// ============ DocumentEtudiant ============
export interface DocumentEtudiant {
  id: number;
  etudiant_id: number;
  type_document: string;
  libelle?: string;
  numero_document?: string;
  date_delivrance?: string;
  lieu_delivrance?: string;
  fichier_url?: string;
  format_fichier?: string;
  taille_fichier?: number;
  statut?: string;
  commentaire?: string;
  is_active?: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface CreateDocumentEtudiant {
  etudiant_id: number;
  type_document: string;
  libelle?: string;
  numero_document?: string;
  date_delivrance?: string;
  lieu_delivrance?: string;
  fichier_url?: string;
  format_fichier?: string;
  taille_fichier?: number;
}

export type UpdateDocumentEtudiant = Partial<Omit<CreateDocumentEtudiant, 'etudiant_id'>>;

// ============ Inscription ============
export interface Inscription {
  id: number;
  etudiant_id: number;
  filiere_id?: number;
  niveau_id?: number;
  annee_academique: string;
  date_inscription?: string;
  type_inscription?: string;
  regime_etudes?: string;
  statut_inscription?: string;
  frais_inscription?: number;
  frais_payes?: number;
  is_active?: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface CreateInscription {
  etudiant_id: number;
  filiere_id?: number;
  niveau_id?: number;
  annee_academique: string;
  date_inscription?: string;
  type_inscription?: string;
  regime_etudes?: string;
  frais_inscription?: number;
  frais_payes?: number;
}

export type UpdateInscription = Partial<Omit<CreateInscription, 'etudiant_id'>>;

export interface InscriptionWithMatieres extends Inscription {
  inscriptions_matieres: InscriptionMatiere[];
}

export interface InscriptionStatistiques {
  total: number;
  par_filiere: Record<string, number>;
  par_niveau: Record<string, number>;
  par_statut: Record<string, number>;
}

// ============ InscriptionMatiere ============
export interface InscriptionMatiere {
  id: number;
  inscription_id: number;
  matiere_id: number;
  semestre: number;
  is_active?: boolean;
  created_at?: string;
}

export interface CreateInscriptionMatiere {
  inscription_id: number;
  matiere_id: number;
  semestre: number;
}

export interface BulkCreateInscriptionMatiere {
  inscription_id: number;
  matiere_ids: number[];
  semestre: number;
}

// ============ Constantes ============
export const STATUTS_ETUDIANT = ['actif', 'suspendu', 'diplômé', 'exclu'] as const;
export type StatutEtudiant = typeof STATUTS_ETUDIANT[number];

export const STATUTS_DOCUMENT = ['en_attente', 'valide', 'refuse'] as const;
export type StatutDocument = typeof STATUTS_DOCUMENT[number];

export const STATUTS_INSCRIPTION = ['en_cours', 'validee', 'annulee'] as const;
export type StatutInscription = typeof STATUTS_INSCRIPTION[number];

export const TYPES_DOCUMENT = [
  'Acte de naissance',
  'Certificat de nationalité',
  'Baccalauréat',
  'Relevé de notes',
  'Photo d\'identité',
  'Carte d\'identité',
  'Passeport',
  'Certificat médical',
  'Autre',
] as const;

export const TYPES_INSCRIPTION = ['nouvelle', 'redoublement', 'transfert'] as const;
export const REGIMES_ETUDES = ['présentiel', 'distance'] as const;

// ============ Couleurs des statuts ============
export const STATUT_COLORS: Record<string, 'success' | 'warning' | 'info' | 'error' | 'default'> = {
  actif: 'success',
  suspendu: 'warning',
  'diplômé': 'info',
  diplome: 'info',
  exclu: 'error',
  en_attente: 'warning',
  valide: 'success',
  refuse: 'error',
  en_cours: 'warning',
  validee: 'success',
  annulee: 'error',
};

File: backend/app/schemas/etudiant.py
"""
Schémas Pydantic pour l'entité Étudiant
"""

from datetime import date, datetime
from typing import Optional, List

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.validators import OptionalEmailEtablissement


class EtudiantBase(BaseModel):
    """Schéma de base pour l'étudiant."""

    nom: str = Field(..., min_length=1, max_length=255, description="Nom de l'étudiant")
    prenom: str = Field(..., min_length=1, max_length=255, description="Prénom de l'étudiant")
    date_naissance: Optional[str] = Field(None, description="Date de naissance (format YYYY-MM-DD)")
    lieu_naissance: Optional[str] = Field(None, max_length=255)
    sexe: Optional[str] = Field(None, max_length=255, description="Sexe (M/F ou MASCULIN/FEMININ)")
    nationalite: Optional[str] = Field(None, max_length=255)
    email: OptionalEmailEtablissement = Field(None, description="Email de l'étudiant")
    telephone: Optional[str] = Field(None, max_length=255)
    telephone_urgence: Optional[str] = Field(None, max_length=50)
    adresse: Optional[str] = Field(None, max_length=500)
    ville: Optional[str] = Field(None, max_length=100)
    pays: Optional[str] = Field("Burkina Faso", max_length=100)
    
    # Informations familiales
    nom_pere: Optional[str] = Field(None, max_length=100)
    profession_pere: Optional[str] = Field(None, max_length=100)
    nom_mere: Optional[str] = Field(None, max_length=100)
    profession_mere: Optional[str] = Field(None, max_length=100)
    personne_contact: Optional[str] = Field(None, max_length=100)
    telephone_contact: Optional[str] = Field(None, max_length=50)
    
    # Informations académiques
    annee_bac: Optional[str] = Field(None, max_length=255)
    numero_bac: Optional[str] = Field(None, max_length=255)
    mention: Optional[str] = Field(None, max_length=255)
    diplome: Optional[str] = Field(None, max_length=255)
    boursier: Optional[bool] = Field(None)
    cni: Optional[str] = Field(None, max_length=255)
    
    # Statut
    statut: Optional[str] = Field("actif", max_length=50)
    is_active: Optional[bool] = Field(True)


class EtudiantCreate(EtudiantBase):
    """Schéma pour la création d'un étudiant (matricule généré automatiquement)."""
    pass


class EtudiantUpdate(BaseModel):
    """Schéma pour la mise à jour d'un étudiant (tous les champs optionnels)."""

    nom: Optional[str] = Field(None, min_length=1, max_length=255)
    prenom: Optional[str] = Field(None, min_length=1, max_length=255)
    date_naissance: Optional[str] = Field(None)
    lieu_naissance: Optional[str] = Field(None, max_length=255)
    sexe: Optional[str] = Field(None, max_length=255)
    nationalite: Optional[str] = Field(None, max_length=255)
    email: OptionalEmailEtablissement = None
    telephone: Optional[str] = Field(None, max_length=255)
    telephone_urgence: Optional[str] = Field(None, max_length=50)
    adresse: Optional[str] = Field(None, max_length=500)
    ville: Optional[str] = Field(None, max_length=100)
    pays: Optional[str] = Field(None, max_length=100)
    nom_pere: Optional[str] = Field(None, max_length=100)
    profession_pere: Optional[str] = Field(None, max_length=100)
    nom_mere: Optional[str] = Field(None, max_length=100)
    profession_mere: Optional[str] = Field(None, max_length=100)
    personne_contact: Optional[str] = Field(None, max_length=100)
    telephone_contact: Optional[str] = Field(None, max_length=50)
    annee_bac: Optional[str] = Field(None, max_length=255)
    numero_bac: Optional[str] = Field(None, max_length=255)
    mention: Optional[str] = Field(None, max_length=255)
    diplome: Optional[str] = Field(None, max_length=255)
    boursier: Optional[bool] = None
    cni: Optional[str] = Field(None, max_length=255)
    numero_carte: Optional[str] = Field(None, max_length=50)
    ine: Optional[str] = Field(None, max_length=50)
    photo_url: Optional[str] = Field(None, max_length=500)
    statut: Optional[str] = Field(None, max_length=50)
    is_active: Optional[bool] = None


class EtudiantInDB(EtudiantBase):
    """Schéma pour un étudiant en base de données."""

    id: int
    matricule: Optional[str] = None
    numero_carte: Optional[str] = None
    ine: Optional[str] = None
    photo_url: Optional[str] = None
    created_by: Optional[str] = None
    created_date: Optional[datetime] = None
    last_modified_by: Optional[str] = None
    last_modified_date: Optional[datetime] = None
    importation_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class Etudiant(EtudiantInDB):
    """Schéma de réponse pour un étudiant."""
    pass


class EtudiantWithDetails(Etudiant):
    """Schéma de réponse pour un étudiant avec ses documents et inscriptions."""

    documents: List["DocumentEtudiantResponse"] = []
    inscriptions: List["InscriptionResponse"] = []


# Import différé pour éviter les imports circulaires
from app.schemas.document_etudiant import DocumentEtudiant as DocumentEtudiantResponse
from app.schemas.inscription import Inscription as InscriptionResponse

EtudiantWithDetails.model_rebuild()
